docs/design/dp_network_policy.md

```diff
@@ -0,0 +1,216 @@
+     return XDP_PASS
+   }
+   
+   if connection_tracking_is_reply(packet) {
+     return XDP_PASS
+   }
+
+   dip_prim_rule = Dest-IP-range-primary-table[packet.VNI, packet.source-ip, packet.dest-ip];
+   dip_supp_rules = Dest-IP-range-supplementary-table[packet.VNI, packet.source-ip, packet.dest-ip];
+   dip_except_rules = dest-ip-range-excpet-table[packet.VNI, packet.source-ip, packet.dest-ip];
+   pp_l4_rules = Proto-port[packet.VNI, packet.source-ip, packet.dest-port, packet.protocol];
+   pp_l3_rules = Proto-port[packet.VNI, packet.source-ip, 0, 0 ];
+
+
+   //perform the bit operation to the get final result, to see if there is any policy allows the packet
+   result =  (dip_prim_rule | (dip_supp_rules & ~dip_except_rules)) & ( pp_l4_rules  | pp_l4_rules )
+   If result != 0 {
+	  return XDP_PASS
+   } else {
+	  return XDP_DROP
+   }
+}
+```
+
+
+### 2. Connection Tracking
+
+#### 2.1 Basic functionality
+
+Connection Tracking is a map that records all active TCP/UDP/SCTP connections. When a connection is established on one direction (either ingress or egress), it's saved into the connection tracking map and then the packets on the other direction will be regarded as reply packets and automatically allowed.
+
+Note: UDP protocol doesn't have a connection concept. Here we are referring to a logic connection identified by a 5-tuple (local IP, local port, remote IP, remote port, UDP).
+
+Below is how connection tracking entires are maintained and checked when inspecting a packet. The same algorithm applies to both ingress traffic and egress traffic:
+ 
+
+```C
+int enforce_policy_on_packet(packet) {
+   if (packet.potocol != UDP|TCP|SCTP) {
+      return XDP_PASS
+   }
+   
+   if connection_tracking_is_reply(packet) {
+      return XDP_PASS 
+   }
+
+   if enforced_by_policy && policy_check_result == false {
+       return XDP_DROP
+   }
+
+    //now we will return pass
+    if connection is not in conn_track {
+       update_connection_tracking(packet)
+    }
+   
+    return XDP_PASS
+}
+```
+
+#### 2.2 Invalidating entries (TBD)
+
+### 3. Label-based Enforcement(TBD)
+
+## Component Changes
+
+### NetworkPolicy Controller
+
+* A new controller, NetworkPolicy controller, will be introduced in this design to watch network policy object CRUD events. 
+* This controller converts the policy objects to low-level BPF map rules, and call daemon to update the rules.
+
+### Pod Controller
+
+* When a pod gets created, updated or deleted, it will check whether the pod is related to existing network policies. 
+* Trigger the policy rule CRUD process if the pod involves network policies.
+
+
+### Mizar Daemon
+
+* New RPC interfaces are implemented to receive the network policy CRUD messages from Mizar master, and then updates the corresponding BPF maps in the kernel space. 
+* Delete invalid entries in connection tracking based on the updated policies.
+
+
+### Transit XDP
+
+* Call the connection tracking module to:
```

Comment:
suggest to re-format this block; it seems the invokation of connection track nodule is the only controlling factor by current formating.

Reply:
Updated.